lib/linkerd2-reconnect/src/service.rs

```diff
@@ -0,0 +1,152 @@
+        }
+    }
+}
+
+impl<Req, T, R, M, S> tower::Service<Req> for Service<T, R, M>
+where
+    T: Clone,
+    R: Recover,
+    M: tower::Service<T, Response = S>,
+    M::Error: Into<Error>,
+    S: tower::Service<Req>,
+    S::Error: Into<Error>,
+{
+    type Response = S::Response;
+    type Error = Error;
+    type Future = future::MapErr<S::Future, fn(S::Error) -> Error>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        loop {
+            self.state = match self.state {
+                State::Disconnected { ref mut backoff } => {
+                    // When disconnected, try to build a new inner sevice. If
+                    // this call fails, we must not recover, since
+                    // `make_service` is now unusable.
+                    try_ready!(self.make_service.poll_ready().map_err(Into::into));
+                    State::Pending {
+                        future: self.make_service.call(self.target.clone()),
+                        backoff: backoff.take(),
+                    }
+                }
+
+                State::Pending {
+                    ref mut future,
+                    ref mut backoff,
+                } => match future.poll() {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(service)) => State::Service(service),
+                    Err(e) => {
+                        // If the service cannot be built, try to recover using
+                        // the existing backoff.
+                        let error: Error = e.into();
+                        tracing::debug!(message="Failed to connect", %error);
+                        State::Recover {
+                            error: Some(error),
+                            backoff: backoff.take(),
+                        }
+                    }
+                },
+
+                State::Service(ref mut service) => match service.poll_ready() {
+                    Ok(ready) => return Ok(ready),
+                    Err(e) => {
+                        // If the service fails, try to recover.
+                        let error: Error = e.into();
+                        tracing::warn!(message="Lost connected service", %error);
+                        State::Recover {
+                            error: Some(error),
+                            backoff: None,
+                        }
+                    }
+                },
+
+                State::Recover {
+                    ref mut error,
+                    ref mut backoff,
+                } => {
+                    // Apply the recovery strategy. If the error isn't fatal,
+                    // prefer the existing backoff to the new one.
+                    let error = error.take().expect("error must be set");
+                    tracing::debug!(message="Service failed", %error);
+                    let new_backoff = self.recover.recover(error)?;
+                    tracing::debug!("Recovering");
```

Comment:
The reconnect module can't know this. It could be logged elsewhere, where it's knowable. (A given backoff impl may not be tied to a fixed duration, etc. Even this implementation doesn't know what it's delay will be until it actually starts doing work).